upgrade_versions.py

```diff
@@ -82,33 +85,142 @@ def rpm_cmp_versions(v1, v2):
+
+        self._bump_release(pkg, changelog)
+
+    def _update_version(self):
+        LOG.info("%s: Updating version to: %s" % (self.pkg,
+                                                  self._repo_version))
+
+        with open(self.pkg.specfile, 'r+') as f:
+            content = f.read()
+
+            version = re.search(r'Version:\s*(\S+)', content).group(1)
+
+            # we accept the Version tag in the format: xxx or %{xxx},
+            # but not: xxx%{xxx}, %{xxx}xxx or %{xxx}%{xxx} because
+            # there is no reliable way of knowing what the macro
+            # represents in these cases.
+            if re.match(r'(.+%{.*}|%{.*}.+)', version):
+                raise exception.PackageSpecError("Failed to parse spec file "
+                                                 "'Version' tag")
+
+            if "%{" in version:
+                macro_name = version[2:-1]
+                content = re.sub(r'(%define\s+%s\s+)\S+' % macro_name,
+                                 r'\g<1>' + self._repo_version, content)
+            else:
+                content = re.sub(r'(Version:\s*)\S+',
+                                 r'\g<1>' + self._repo_version, content)
+
+            # since the version was updated, set the Release to 0. When
+            # the release bump is made, it will increment to 1.
+            content = re.sub(r'(Release:\s*)[\w.-]+', r'\g<1>0', content)
+
+            f.seek(0)
+            f.write(content)
+
+    def _bump_release(self, pkg, log=None):
+        LOG.info("%s: Bumping release" % self.pkg)
 
-    def bump_release(self):
         if self.pkg.commit_id:
-            pkg = copy.copy(self.pkg)
-            pkg.commit_id = None
-            pkg.download_source_code()
-            pkg.commit_id = pkg.repository.repo.head.target.hex[:7]
             if pkg.commit_id != self.pkg.commit_id:
-                print("Updating package %s from %s to %s" % (
-                      self.pkg.name, self.pkg.commit_id, pkg.commit_id))
+                LOG.info("Updating package %s from %s to %s" % (
+                    self.pkg.name, self.pkg.commit_id, pkg.commit_id))
                 log = _get_git_log(pkg.repository.repo, self.pkg.commit_id)
-                rpm_bump_spec(pkg.specfile, log)
                 _sed_yaml_descriptor(self.pkg.package_file, self.pkg.commit_id,
                                      pkg.commit_id)
 
+        if log:
+            rpm_bump_spec(pkg.specfile, log)
+
+    def _read_spec(self):
+        self._spec_version = rpm_query_spec_file('version', self.pkg.specfile)
+        LOG.info("%s: Current version: %s" % (self.pkg, self._spec_version))
+        self._spec_release = rpm_query_spec_file(
+            'release', self.pkg.specfile).split('.')[0]
+
+    def _read_version_from_repo(self, repo_path):
+
+        version_regexes = [
+            self.pkg.version_file_regex,
+            ('VERSION', r'.*'),
+            ('%s.spec' % self.pkg, r'Version:(.*)'),
```

Comment:
That was on purpose because I'm using `str.strip()` later on, but I guess there's no harm in changing it.